### DISCUSSION TITLE
What function does C-c C-c call?

### DISCUSSION THREAD

**Discussion by aykaramba** [lem-project/lem#1331] [General]
I need to create a new shortcut that does the following:

* Compile the local lisp function with whatever C-c C-c calls.
* Call my function that does something else.

I am using the default start-lisp-repl, not sure what that is, sly, slime?

Thanks.

**Comment by aykaramba** [comment]
Looks like its (lisp-compile-defun).

**Comment by aykaramba** [comment]
Okay, so this works in the Lem config:

```
(define-key *global-keymap* "C-c C-v" 'clog-stuff)

(define-command clog-stuff ()
    (lem-lisp-mode:lisp-compile-defun)
  )
```

However, let's say I am working on a particular package I created via quicklisp and it is located in ~/common-lisp/mypackage.  Let's say that package has a (myfunction () ..) that does something and let's say that I want to call it after the lisp-compile-defun in the agove (define-command ...).  On the repl i would (in-package :mypackage) then (myfunction ()) and it works.  How would this be best setup in Lem?

If I create: 

```
(define-command clog-stuff ()
    (lem-lisp-mode:lisp-compile-defun)
    (mypackage:myfunctions())
  )
```

When Lem starts up there is an error because my package is not loaded up yet.  I think I want to just start Lem so that the command is "available to be called" but not until C-c C-v is invoked and it should work after I manually decide to (ql:quickload :mypackage).

Any suggestions on how to write that (define-command)?  Do I need to use a (lambda () ... ) for this?

**Reply by vindarel** [reply]
For this kind of stuff we can use `uiop:symbol-call package function` and also use the `*features*` flags: `push` a feature to the list when you quickload your package, and check its presence with `#+my-feature`.

**Reply by aykaramba** [reply]
Aha!  Thank you, I will look that up.  I am taking your Lisp course btw, thanks for all of that.

**Reply by vindarel** [reply]
oooh thank you. I hope it serves you well.

**Comment by garlic0x1** [comment]
> I am using the default start-lisp-repl, not sure what that is, sly, slime?

That command gives you a REPL to Lem itself, you usually want `M-x slime` for working on separate projects.

If you didn't already figure it out, you can use `M-x describe-key` to figure out what command a key is bound to, then use `M-x apropos-command` to find info about that command and what source file it is in.

> Any suggestions on how to write that (define-command)? Do I need to use a (lambda () ... ) for this?

First off, check other usages of `define-command` to see the syntax used.  You are missing the list for prompts after the argument list.  
If you want it to be loaded in Lem at startup you should put custom commands in your Lem config.

**Comment by aykaramba** [comment]
Thank you kindly for the info garclic0x1.  Adding that to my todo list.

**Comment by aykaramba** [comment]
Just wanted to follow up in case some user bumps into this thread.   Here is how you can create a shortcut to compile a form and do some extra things at the same time.  In my case:

1) I am working in CLOG to create my-app in CLOG.

2) Want to be able to compile forms and reload the CLOG web page so I put the following function into the main (defun on-new-window () ..) form as follows:

```
(defun reload (obj)
  (url-replace (location body) "/"))
```

3) We can now create a new key binding to be able to compile a form and then call the (reload ()) function to reload the CLOG page:

```
(in-package :lem-user)
(ql:quickload :my-app)
(define-key lem-lisp-mode:*lisp-mode-keymap* "C-c C-v" 'clog-reload)
(define-command clog-reload () ()
      (my-app::reload 123)
  )
``` 

4) For me this reduces the number of round trips to reload the page each time I try something new.  It also opens up an insight into how we can start to use LEM and get additional interactivity while developing in CLOG. 

Thanks for the help everyone!

LEM + Lisp are pure awesome.

**Comment by aykaramba** [comment]
Dropping in to update this with a new config to help other noobs with Lem and CLOG:

1) I started off by customizing Lem via the init file.  However, because I use Lem everywhere I did not want to load up my clog project every time and I did not want to figure out how to manage custom config files as that is not documented yet.

2) Besides, the customization are specific to the project and specific to the fact it is a CLOG app and does not apply to other types of projects I might want to build.

3) So, I made the above local to my CLOG app as follows:

a) Put the customizations in a command.lisp file in my CLOG project folder:

```
(in-package :lem-user)
(ql:quickload :my-app)
(define-key lem-lisp-mode:*lisp-mode-keymap* "C-c C-v" 'clog-reload)
(define-command clog-reload () ()
      (my-app::reload 123)
  )
```

b) Added the file to my :components section in the my-app.asd file:
```
  :components ((:file "my-app")
                            (:file "commands")))
```

4) Now when I quickload my-app the customizations are loaded up as well, super convenient.

5) The downside of this setup is that if I wanted to share the source with someone else and they weren't using lem, the user would get an error.  So, the next step would be to learn how to do something like this in a way that allows for conditional loading of the customizations based on project.  Perhaps this might be better handled by the Lem "projects" feature, but that is something to review in the future.
